src/transport/SessionManager.cpp

```diff
@@ -573,16 +582,30 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
         return;
     }
 
-    // Trial decryption with GroupDataProvider. TODO: Implement the GroupDataProvider Class
-    // TODO retrieve also the fabricIndex with the GroupDataProvider.
-    // VerifyOrExit(CHIP_NO_ERROR == groups->DecryptMessage(packetHeader, payloadHeader, msg),
-    //     ChipLogError(Inet, "Secure transport received group message, but failed to decode it, discarding"));
+    Credentials::GroupDataProvider::GroupSession groupContext;
+    if (packetHeader.IsValidGroupMsg())
+    {
+        // Trial decryption with GroupDataProvider
+        auto iter = groups->IterateGroupSessions(packetHeader.GetSessionId());
+        VerifyOrReturn(nullptr != iter);
 
-    ReturnOnFailure(payloadHeader.DecodeAndConsume(msg));
+        System::PacketBufferHandle msgCopy;
+        bool decrypted = false;
+        while (!decrypted && iter->Next(groupContext))
+        {
+            msgCopy   = msg.CloneData();
+            decrypted = (CHIP_NO_ERROR ==
+                         SecureMessageCodec::Decrypt(CryptoContext(groupContext.key), payloadHeader, packetHeader, msgCopy));
+        }
+        iter->Release();
+        VerifyOrReturn(decrypted);
+        msg = msgCopy.CloneData();
```

Comment:
I think `msg = std::move(msgCopy)` would work without an additional copy

Reply:
Changed as suggested. Need to re-test, but I think it'll work.